pilot/platform/kube/inject/file.go

```diff
@@ -0,0 +1,170 @@
+
+package inject
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"reflect"
+
+	"github.com/ghodss/yaml"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	// TODO(nmittler): Remove this
+	_ "github.com/golang/glog"
+
+	appsv1beta1 "k8s.io/api/apps/v1beta1"
+	batchv1 "k8s.io/api/batch/v1"
+	"k8s.io/api/batch/v2alpha1"
+	"k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	yamlDecoder "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+var (
+	kinds = []struct {
+		groupVersion schema.GroupVersion
+		obj          runtime.Object
+		resource     string
+		apiPath      string
+	}{
+		{v1.SchemeGroupVersion, &v1.ReplicationController{}, "replicationcontrollers", "/api"},
+
+		{v1beta1.SchemeGroupVersion, &v1beta1.Deployment{}, "deployments", "/apis"},
+		{v1beta1.SchemeGroupVersion, &v1beta1.DaemonSet{}, "daemonsets", "/apis"},
+		{v1beta1.SchemeGroupVersion, &v1beta1.ReplicaSet{}, "replicasets", "/apis"},
+
+		{batchv1.SchemeGroupVersion, &batchv1.Job{}, "jobs", "/apis"},
+		{v2alpha1.SchemeGroupVersion, &v2alpha1.CronJob{}, "cronjobs", "/apis"},
+		// TODO JobTemplate requires different reflection logic to populate the PodTemplateSpec
+
+		{appsv1beta1.SchemeGroupVersion, &appsv1beta1.StatefulSet{}, "statefulsets", "/apis"},
+	}
+	injectScheme = runtime.NewScheme()
+)
+
+func init() {
+	for _, kind := range kinds {
+		injectScheme.AddKnownTypes(kind.groupVersion, kind.obj)
+		injectScheme.AddUnversionedTypes(kind.groupVersion, kind.obj)
+	}
+}
+
+func intoObject(sidecarTemplate string, meshconfig *meshconfig.MeshConfig, in runtime.Object) (interface{}, error) {
+	out := in.DeepCopyObject()
+
+	var metadata *metav1.ObjectMeta
+	var podSpec *v1.PodSpec
+
+	// CronJobs have JobTemplates in them, instead of Templates, so we
+	// special case them.
+	if job, ok := out.(*v2alpha1.CronJob); ok {
+		metadata = &job.Spec.JobTemplate.ObjectMeta
+		podSpec = &job.Spec.JobTemplate.Spec.Template.Spec
+	} else {
+		// `in` is a pointer to an Object. Dereference it.
+		outValue := reflect.ValueOf(out).Elem()
+
+		templateValue := outValue.FieldByName("Spec").FieldByName("Template")
+		// `Template` is defined as a pointer in some older API
+		// definitions, e.g. ReplicationController
+		if templateValue.Kind() == reflect.Ptr {
+			templateValue = templateValue.Elem()
+		}
+		metadata = templateValue.FieldByName("ObjectMeta").Addr().Interface().(*metav1.ObjectMeta)
+		podSpec = templateValue.FieldByName("Spec").Addr().Interface().(*v1.PodSpec)
+	}
+
+	// Skip injection when host networking is enabled. The problem is
+	// that the iptable changes are assumed to be within the pod when,
+	// in fact, they are changing the routing at the host level. This
+	// often results in routing failures within a node which can
+	// affect the network provider within the cluster causing
+	// additional pod failures.
+	if podSpec.HostNetwork {
```

Comment:
Should we log when we skip injection? A debug level log would be sufficient for troubleshooting these subtle cases.